rclcpp/include/rclcpp/timer.hpp

```diff
@@ -133,6 +135,13 @@ class GenericTimer : public TimerBase
   )
   : TimerBase(clock, period, context), callback_(std::forward<FunctorT>(callback))
   {
+    TRACEPOINT(
+      rclcpp_timer_callback_added,
+      (const void *)get_timer_handle().get(),
+      (const void *)&callback_);
+    // TODO(christophebedard) maybe there's a simpler way
+    // to treat callback_ as an std::function here explicitly
```

Comment:
@iluetkeb not following this TODO, mind to explain?

Reply:
I'm not @iluetkeb :laughing:, but IIRC it was about finding a way to avoid needing any `register_callback_function()` methods at all (with `std::enable_if`). It seemed to be necessary, because moving

```
TRACEPOINT(
  rclcpp_callback_register,
  (const void *)&callback_,
  get_symbol(function));
```

to the constructor, instead of calling `this->register_callback_function()` and then casting it to its actual type, does not work. I didn't know how to avoid this.

Reply:
Do you happen to know why it does not work?

Reply:
Not unfortunately. I'd defer to people more knowledgeable about C++ templating. It seems like it just does not know what type it actually is in the constructor, and that I need to use `std::enable_if<>` to figure that out.

Reply:
I could more into it, though

Reply:
Did we make any progress on this? Going over this again, and without any additional context. it seems you are forced to cast the provided `FunctorT` callback to a concrete callback type to retrieve its symbol. If you guys used SFINAE on `get_symbol` I think you'd be able to put the macro back in the constructor.

Reply:
To be honest, while I think I understand SFINAE, I've played around a bit and I'm not sure how I'm supposed to use apply it to `get_symbol()`.

I'm not sure if forward-declaring `rclcpp::TimerBase` in `tracetools` (which is what I'm assuming I would have to do?) really makes this worth being able to move the macro to the constructor.

Reply:
> I'm not sure if forward-declaring `rclcpp::TimerBase` in `tracetools`

In principle, you shouldn't need to do that. And actually, since you know the exact types there's no need for template metaprogramming, just locally overload `get_symbol` appropriately and [ADL](https://en.cppreference.com/w/cpp/language/adl) will do the rest. 

I'm curious about https://github.com/ros2/rclcpp/pull/789#discussion_r320478335 and why/how it did not work though.

Reply:
Yeah, I've come to the same conclusion, and I'll try that first.

Reply:
@hidmic @iluetkeb solved! The reason why it wouldn't compile with the call to `get_symbol` directly in the constructor was because the tests use lambdas with captures. Apparently they're different and can't really be `std::function`s. Not sure why the `std::enable_if`s work though, since they only check for `std::function`s.

The solution is just to add a `get_symbol` overload which takes a generic 
type and calls `typeid()`, which is exactly what we want for lambdas.

I think this makes more sense, since `tracetools` isn't actually aware of this intricacy in `rclcpp` and simply provides a generic fallback.

Reply:
The PR has been updated and I consider this resolved, then. Plz reopen otherwise.